### DISCUSSION TITLE
OBS_LEAD and pcp_combine/gridstat

### DISCUSSION THREAD

**Discussion by willmayfield** [dtcenter/METplus#2318] [Configuration] [answered]
Hello,

I was recently using gridstat with ccpa accumulated precip data, and noticed the following behavior:

- When using ccpa data, which comes in 1-h buckets of accumulated precip in grib2 files, directly into gridstat, the output .stat files have the OBS_LEAD column filled with values "010000".

- If I first run pcp_combine with these data, creating the met netcdf formated observations, _then_ use them with gridstat, OBS_LEAD is filled with the value "000000."

- Similarly, running grid_stat with nohrsc accumulated snowfall grib2 files, which are available at 6-h accumulation buckets, OBS_LEAD has value "060000". But if pcp_combine is run first, OBS_LEAD becomes 000000.

I wasn't able to track down exactly why this happens, but it may be coming from the differing ways that the data_plane method loads data from grib2 vs met-netcdf formats.

While we don't have a use for the OBS_LEAD column in our application, we thought it might be worth bringing up in case there was an intention for OBS_LEAD to contain the accumulation length. (But I'm undecided if that even makes sense, and have also noted that specific to MODE applications there is an "OBS_ACCUM" value). 

Cheers!

Will

**Comment by JohnHalleyGotway** [comment, marked as answer]
Hi @willmayfield, I see you have a question about metadata. You notice that the logic for handling the "lead time" for CCPA data seems inconsistent. I definitely agree that metadata is important to consider. It's often the case that it just "is what it is" and we have relatively little control over it. Let me describe what's happening in the code, and then we can decide if the apparent inconsistency warrants any code changes.

1. Pull two sample GRIB2 CCPA data files containing 1-hour precip:
```
wget https://ftp.ncep.noaa.gov/data/nccf/com/ccpa/prod/ccpa.20230820/12/ccpa.t11z.01h.1p0.conus.gb2
wget https://ftp.ncep.noaa.gov/data/nccf/com/ccpa/prod/ccpa.20230820/12/ccpa.t12z.01h.1p0.conus.gb2
```
2. Run `plot_data_plane` at verbosity level 4 to print out metadata info:
```
plot_data_plane ccpa.t12z.01h.1p0.conus.gb2 ccpa.ps 'name="APCP"; level="A1";' -v 4
...
DEBUG 4: Data plane information:
DEBUG 4:    plane min: 0
DEBUG 4:    plane max: 4.90344
DEBUG 4:    valid time: 20230820_120000
DEBUG 4:    lead time: 010000
DEBUG 4:    init time: 20230820_110000
DEBUG 4:    accum time: 010000
```
When reading data from GRIB2 files, the lead time is computed on this [line of code](https://github.com/dtcenter/MET/blob/266a9bdc54e2881851056f179dd42fcdd07c667e/src/libcode/vx_data2d_grib2/data2d_grib2.cc#L841) as the valid time minus the initialization time. And you see that reflected in the log messages above.
3. Run `pcp_combine` to add 1-hour precip with a *single input file*.
```
pcp_combine -add ccpa.t12z.01h.1p0.conus.gb2 -field 'name="APCP"; level="A1";' add1.nc -v 4
```
Check the metadata in the NetCDF output file and the 1-hour init/valid time difference exists:
```
ncdump -h add1.nc
...
	float APCP_01(lat, lon) ;
		APCP_01:init_time = "20230820_110000" ;
		APCP_01:valid_time = "20230820_120000" ;
		APCP_01:accum_time = "010000" ;
```
4. Next, add data from *two input files* and check the metadata in the NetCDF output file.
```
pcp_combine -add ccpa.t11z.01h.1p0.conus.gb2 ccpa.t12z.01h.1p0.conus.gb2 \
-field 'name="APCP"; level="A1";' add2.nc -v 4
ncdump -h add2.nc
...
	float APCP_02(lat, lon) ;
		APCP_02:init_time = "20230820_120000" ;
		APCP_02:valid_time = "20230820_120000" ;
		APCP_02:accum_time = "020000" ;
```
Here's the punchline... while the accumulation time indicates 2 hours of accumulation, the initialization and valid times are set to the SAME TIME. They do not differ by 2 hours! When this data is read into `grid_stat`, the lead time is computed as valid minus init and that yields a value of 0... rather than 1 or 2.

Looking in `pcp_combine.cc` at the derivation logic (which is used for `-add` option):
- This [line of code](https://github.com/dtcenter/MET/blob/266a9bdc54e2881851056f179dd42fcdd07c667e/src/tools/core/pcp_combine/pcp_combine.cc#L978), initializes the output initialization time to the init time of the first field it reads.
```
nc_init_time  = cur_dp.init();
```
- Next, starting on this [line of code](https://github.com/dtcenter/MET/blob/266a9bdc54e2881851056f179dd42fcdd07c667e/src/tools/core/pcp_combine/pcp_combine.cc#L1014), it checks to see if the init time *CHANGES* in subsequent files read. If so, it resets it to a default value of 0. For the valid time, it keeps track of the *MAXIMUM* valid time encountered.
- Last, this [line of code](https://github.com/dtcenter/MET/blob/266a9bdc54e2881851056f179dd42fcdd07c667e/src/tools/core/pcp_combine/pcp_combine.cc#L1417), resets the init time value of 0 to the valid time.

So what *should* the logic in `pcp_combine` be? The assumption is that when you combine data with the same initialization time, it's from the same model run, and that init time should be preserved and reported in the output. But when the init times differ, then they are NOT from the same model run. We could report the minimum init time, but that'd imply that the input data IS from the same model run... which is misleading. With no obvious better choice, we chose to set init = valid.

I realize that leads to the apparent inconsistency you note. But it is not obvious to me that any of this logic should be changed.

I'll also note that an option for overriding this does exist using the `set_attr_init` config option:
```
pcp_combine -add ccpa.t11z.01h.1p0.conus.gb2 ccpa.t12z.01h.1p0.conus.gb2 \
-field 'name="APCP"; level="A1"; set_attr_init="20230820_10";' add2.nc -v 4
```
In this example, `set_attr_init="20230820_10";` overrides the initialization time read from all input data and sets it to the specified time. Since that init time remains constant for all files `pcp_combine` reads, it will reported to the ouput file:
```
	float APCP_02(lat, lon) ;
		APCP_02:init_time = "20230820_100000" ;
		APCP_02:valid_time = "20230820_120000" ;
```
So the workaround exists, albeit a little clunky.

**Reply by willmayfield** [reply]
@JohnHalleyGotway Thank you or your thorough reply! This makes sense and thank you for explaining the workaround you gave here. The logic you explained makes perfect sense for how the different initialization times mean different model runs for forecasts---I guess it isn't surprising that observation buckets don't really fit with the same vocabulary, so we will keep this in mind.

Thanks!

Will
